fix(array-virtual-repeat-strategy): fix mutation handling

bigopon opened the pull request:
Thanks to @reinholdk , our array strategy is much more resilience to mutation issues.

About the real issue we see in https://dist-fymprbfvcn.now.sh/#/issue-97 when:

```
- focus on input
- scroll down pass 2 - 3 screen
- type something
```

 what gonna happen is:

```
- screen scrolled back to top (no scroll event)
- input change
- mutate array
```

Unfortunately there seems to be no way to handle the first step easily, so I think we can temporarily make it a caveat that: if by some method/way that a container is scrolled without firing scroll event, we cannot guarantee virtual repeat correctness.

We can test this behavior by programmatically change the value of input and dispatch input event, you will see that it will work just fine.

@reinholdk I made you the authors of the commits, though I don't have your email so GitHub is not showing your icon in those commits. If you want to change that, please let me know. I also adjusted the tests and extract your tests out into `async/await` manner so it can be easily handled without having it randomly fail.

@reinholdk about your fix in `validateScrolledState`, I don't feel it's working, as a few of existing tests started to fail when I added those changes in.

If you want, you can also rebase on master on continue the work on your PR, I'm only doing this as I recently have made a big refactor converting this repo to TS. So let me know whether you are ok with doing it this way.

cc @EisenbergEffect @AStoker 

supersedes #101
fixes #97
should fixes #103

reinholdk commented:
@bigopon thanks for picking this up! Will run some tests on top of your PR. Need some time though.

bigopon commented:
@reinholdk no worries. Note that i've added a few commits that employ your change in test, against mutation + scrolling behavior after that, and they work well. So I'm a bit confused about correctness state of existing tests. For all of your new tests, I have refactored them here https://github.com/aurelia/ui-virtualization/pull/145/files#diff-8694910f0954da8acf19cede258ecec5R125

you can have a look at these tests to see what I'm about to test the new change with, beside your existing new tests https://github.com/aurelia/ui-virtualization/blob/260527f295c98d7df9e30f500a04799e9790882b/test/vr-integration.splicing-2.spec.ts#L81-L93

Edit: If splice with remove only, things work fine, when there is an add, it starts to break. I'm looking for a better refactor overall.

bigopon commented:
Should revisit this later. Parts of this has already been merged in other PRs
